### DISCUSSION TITLE
Measure the size of compressed image

### DISCUSSION THREAD

**Discussion by danishnazir** [InterDigitalInc/CompressAI#215] [Q&A] [answered]
Hi All,

I just wanted to ask how can we get exact file size of the compressed image? For example if an image `x` is passed to a pre-trained model, which produces the latent space at some fixed `bpp`, then how can we measure the size of the compressed image itself? Do we save the actual bits (in the form of strings) produced by the  entropy encoder? Can someone please point me to the right direction?

Regards,

**Comment by YodaEmbedding** [comment, marked as answer]
Two methods, which we use in the given \[Code\] links:

- [\[Code\]](https://github.com/InterDigitalInc/CompressAI/blob/v1.2.4/compressai/losses/rate_distortion.py#L52-L55). During training, we calculate the *negative log likelihoods* via:
   ```python
    bpp_loss = sum( -log2(likelihoods) ) / num_pixels
   ```
   The `likelihoods` representing $p_{\hat{y}}(\hat{y})$ are determined by feeding a monotonic differentiable function $c : \mathbb{R} \to [0, 1]$ (i.e., the CDF) with a single latent tensor element $\hat{y}_i$. Further details can be found in the Appendix of ["Variational image compression with a scale hyperprior"](https://arxiv.org/abs/1802.01436).  

- [\[Code\]](https://github.com/InterDigitalInc/CompressAI/blob/v1.2.4/compressai/utils/eval_model/__main__.py#L124). During inference, when using a real entropy coder,
   ```python
   bpp = len(bytestring) * 8 / num_pixels
   ```

If measured on the exact same $\hat{y}$, the aforementioned `bpp_loss` and `bpp` should be very, very close to each other, for a well-trained model that generates reasonable distributions from reasonable input images.

Please follow the [training tutorial](https://interdigitalinc.github.io/CompressAI/tutorials/tutorial_train.html) if you want to try this out on your own [custom models](https://interdigitalinc.github.io/CompressAI/tutorials/tutorial_custom.html).

---

Related:

- https://github.com/InterDigitalInc/CompressAI/issues/191#issuecomment-1369577829
- https://github.com/InterDigitalInc/CompressAI/issues/199#issuecomment-1418253326

**Reply by RizerKayseri** [reply]
Hello, during the process of using, I encountered something quite confusing. In the case where num_pixels equals 1024, the calculated bpp_loss consistently remains around 0.9. However, for the length of the bitstring (len(bitstring)), its result always approximates to about 120. This means that when I actually compress data, the bpp (bits per pixel) obtained by calculating bpp = len(bitstring) / num_pixels is significantly smaller than the bpp_loss.

**Reply by YodaEmbedding** [reply]
```python
len_bitstring = 8 * len(bytestring)
```

I've updated my earlier post for clarity.

**Reply by RizerKayseri** [reply]
Thank you very much

**Reply by RizerKayseri** [reply]
I made a foolish mistake.

**Comment by danishnazir** [comment]
Thank you for the explaination @YodaEmbedding , just wanted to double check my understanding: The size of `bpp` in Bytes is the actual (during testing ofcourse) filesize on any input `x`?

**Reply by YodaEmbedding** [reply]
The file size in bytes of the compressed image is `bpp * num_pixels / 8`.
